### PR TITLE
Compute screen lines lazily after a reset

### PR DESCRIPTION
**Pull request by maxbrunsfeld**
This is one solution to the problem of resetting a TextEditor's display layer repeatedly when configuring it from the outside (https://github.com/atom/atom/pull/12125).

I believe this brings file-opening performance on my Atom PR back to parity with master. I'm not sure that I love it, but opening for your consideration @as-cii @nathansobo.

### CODE REVIEW

File: src/display-layer.coffee [hunk 1]
```diff
@@ -1212,6 +1223,7 @@ class DisplayLayer
     }
 
   getScreenLineCount: ->
+    @recomputeScreenLinesIfNeeded()
```
Comment (as-cii) [inline src/display-layer.coffee]: Should we add this to methods like `getRightmostScreenPosition` and `lineLengthForScreenRow` as well?

File: src/display-layer.coffee [hunk 2]
```diff
@@ -126,6 +127,9 @@ class DisplayLayer
       "\r\n": @invisibles.cr + @invisibles.eol
     }
 
+  recomputeScreenLinesIfNeeded: ->
```
Comment (as-cii) [inline src/display-layer.coffee]: What about renaming this to `recomputeAllSpatialScreenLinesIfNeeded()`? In this file we usually refer to "screen lines" when talking about the synthesized lines that contain both spatial information and decoration tags.

File: spec/display-layer-spec.coffee [hunk 3]
```diff
@@ -33,6 +34,39 @@ describe "DisplayLayer", ->
       expect(displayLayer2.ratioForCharacter).toBe(displayLayer1.ratioForCharacter)
       expect(displayLayer2.isWrapBoundary).toBe(displayLayer1.isWrapBoundary)
 
+  it "does not recompute the display layer's screen lines until they are needed", ->
+    spliceSpy = spyOn(DisplayIndex.prototype, 'splice').and.callThrough()
+    buffer = new TextBuffer(text: '\tif (a) { b; } else { c; }')
+
+    # construction does not force screen line computation
+    displayLayer = buffer.addDisplayLayer(tabLength: 4)
+    expect(spliceSpy.calls.count()).toBe(0)
```
Comment (as-cii) [inline spec/display-layer-spec.coffee]: I think I am 👍 on spying on `splice`. Just as an alternative, we might consider querying the state of the `DisplayIndex` and avoid the spy:

```js
expect(displayLayer.displayIndex.getScreenLineCount()).toBe(1)
```

File: src/display-layer.coffee [hunk 4]
```diff
@@ -126,6 +127,9 @@ class DisplayLayer
...
+    return unless @propertiesHaveChanged
+    @propertiesHaveChanged = false
```
Comment (as-cii) [inline src/display-layer.coffee]: Super minor stylistic comment here. 🎨 What do you think about changing this into:

```coffee
if @propertiesHaveChanged
  @propertiesHaveChanged = false
  # etc. etc.
```

Feel free to ignore this if you don't think it helps readability. 👍

**Comment by as-cii** [comment]
Left a couple of comments, but overall I think  I am 👍 with this approach. We are going to need something similar anyways if we want to batch buffer changes in the future, so it feels reasonable to me to go down this path.

**Comment by nathansobo** [comment]
What about the problem of delaying the change event after calling `reset`?

**Comment by maxbrunsfeld** [comment]
I think I'm putting this on 🍧  in favor of a different solution in Atom.

File: spec/display-layer-spec.coffee (hunk 3 above)
Comment (maxbrunsfeld) [inline spec/display-layer-spec.coffee]: Good point; that's a much better way of writing the assertion.

**Pull request closed without merging**